versioning/lockfiles/build_order.rst

```diff
@@ -0,0 +1,211 @@
+
+[IMAGE]
+
+.. note::
+
+    The code used in this section, including a *build.py* script to reproduce it, is in the
+    examples repository: https://github.com/conan-io/examples. You can go step by step
+    reproducing this example while reading the below documentation.
+
+    .. code:: bash
+
+        $ git clone https://github.com/conan-io/examples.git
+        $ cd features/lockfiles/build_order
+        # $ python build.py only to run the full example, but better go step by step
+
+
+The example in this section uses ``full_version_mode``, that is, if a package changes any part of its version, its consumers will
+need to build a new binary because a new ``package_id`` will be computed. This example will use version ranges, and
+it is not necessary to have revisions enabled. It also do not require a server, everything can be reproduced locally.
+
+
+.. code-block:: bash
+
+    $ conan config set general.default_package_id_mode=full_version_mode
+
+Let's start by creating the initial dependency graph, without binaries (just the exported recipes), in our local cache:
+
+
+.. code-block:: bash
+
+    $ conan export pkga pkga/0.1@user/testing
+    $ conan export pkgb pkgb/0.1@user/testing
+    $ conan export pkgc pkgc/0.1@user/testing
+    $ conan export pkgd pkgd/0.1@user/testing
+    $ conan export app1 app1/0.1@user/testing
+    $ conan export app2 app2/0.1@user/testing
+
+Now we will create a lockfile that captures the dependency graph for ``app1/0.1@user/testing``.
+In the same way we created lockfiles for a local ``conanfile.py`` in a user folder, we can also
+create a lockfile for a recipe in the Conan cache, with the ``--reference`` argument:
+
+.. code-block:: bash
+
+    $ conan lock create --reference=app1/0.1@user/testing --lockfile-out=app1.lock
+
+The resulting *app1.lock* lockfile will not be able to completely lock the binaries because such
+binaries do not exist at all. This can be checked in the *app1.lock* file, the packages do not
+contain a package revision (``prev``) field at all:
+
+.. code-block:: text
+
+     {
+        ...
+        "4": {
+        "ref": "pkga/0.1@user/testing",
+        "options": "",
+        "package_id": "5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9",
+        "context": "host"
+        }
+        ...
+     }
+
+We can now compute the "build-order" of the dependency graph. The "build-order" lists
+in order all the packages that needs to be built from sources. As the lockfile contains
+the information of existing locked package binaries, the logic is the following:
+
+- If a package is fully locked (it contains a package revision field ``prev`` in the lockfile),
+  it cannot be built from sources, so it will never be returned
+- If a package is not fully locked (it does not contain a package revision ``prev`` in the lockfile),
+  the package will be returned. This situation happens both when the package binary doesn't exist yet,
```

Comment:
```suggestion
- If a package is not fully locked (it does **not** contain a package revision ``prev`` in the lockfile),
  it will appear in the build-order list. This situation happens both when the package binary doesn't exist yet,
```